### ISSUE TITLE
Performance with python build?

### ISSUE THREAD

**Issue by gbrow004** [NVIDIA/apex#355]
Is there some sense on how the python build performs, compared with the C/NCCL compile? I'm not getting an error, but I get the warning :
```
Warning:  multi_tensor_applier fused unscale kernel is unavailable, possibly because apex was installed without --cuda_ext --cpp_ext. Using Python fallback.  Original ImportError was: ModuleNotFoundError("No module named 'amp_C'")
```
So, it's still running. I guess I wanted to know much more performance I would be getting if I chased this down (our institution makes it difficult to troubleshoot sometimes). I have a fairly  large (500GB) run across 6x V100's on our DGX1 on opt_level 01.
Ubuntu 16.04.6
Pytorch 1.1.0
CUDA 10.0

**Comment by mcarilli** [comment]
It depends on your model.  If your model is large, and most of the time is spent in the forward/backward pass, the Python-only build should not be a major drawback.  O0 and O3 ("pure fp32" and "pure 16"** training respectively) with `loss_scale=1.0` should not use the C++ backend at all even if it's available, so these would be negligibly affected.  

To assess how well mixed precision is doing in terms of speed, you can try
`opt_level='O0', loss_scale=1.0` and 
`opt_level=O3, loss_scale=1.0, keep_batchnorm_fp32=True`**
and compare the speed to what you see with 
`opt_level='O1'`.
(for `O1`, do not include `loss_scale=1.0` as an explicit argument, so it can use dynamic loss scaling).

The `O0` options will give you an FP32 speed baseline.  The `O3` options will give you a speed ceiling (and might not converge, but that's not relevant to speed testing).  `O1` should fall somewhere between the two, hopefully towards the fast end.

**Keeping batchnorm in FP32 no matter what can improve performance because it enables cudnn batchnorm.
